### PR TITLE
Feature/mingw fixes

### PR DESCRIPTION
**Pull request by angerman**
###### Motivation for this change

Allow cross compiling haskell programs to windows using mingw

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by domenkozar** [comment]
Is removing lib subfolder required for this work? I saw you missed a couple of instances.

**Comment by angerman** [comment]
@domenkozar to compile `cardano-sl`, yes. Otherwise I run into:
```
x86_64-pc-mingw32-gcc: error trying to exec '/nix/store/g79d8qxa4j8y50sjcinb5j6c54zhq4wg-x86_64-pc-mingw32-stage-final-gcc-debug-7.3.0/libexec/gcc/x86_64-pc-mingw32/7.3.0/collect2': execv: Argument list too long
```

In general I think putting the haskell libs in `lib` is bad, because the underlying logic in nix, expects `$out/lib` folder to contain libraries (such that dependencies, are automatically discovered). For haskell, the libs aren't even directly in the `lib` folder. 

If you could outline which `lib`s I missed, I'd be happy to fix them up :-)

However I know there was some alternative to this fix somewhere(?) using response files instead? I'm slightly tending against such a solution, as it would still unnecessarily polute the library serach paths in nix.

Maybe we need a way to flag some `lib` folders not to be automatically placed into the library search paths. In that case I don't see why we can't just not put them in `lib` in the first place.

**Comment by domenkozar** [comment]
I think the proper solution is to not duplicate so many flags: https://github.com/NixOS/nixpkgs/issues/41340#issuecomment-403453874

**Comment by angerman** [comment]
@domenkozar sure. But why put them in there in the first place? Any haskell tooling will go through the package-db, and pull the proper LD flags anyway. Why have nix provide LD flags what are useless?

E.g. why inject `<pkg>/lib`, when you actually want `<pkg>/lib/$ghc` anyway?

Don't get me wrong. Automatically injecting `$out/lib` is often the right thing to do, when libraries are put into `$out/lib`. But Haskell does for better or worse not really follow that approach, and expects the package-db to be the canonical source for path information.

**Comment by angerman** [comment]
#41340 should of course be fixed as well.

**Comment by domenkozar** [comment]
@angerman I agree it may be changed, but note that it should rather be done in https://github.com/NixOS/nixpkgs/pull/32629 and fix the GCC flag pollution properly 

@angerman I couldn't compile `cachix` on darwin with your lib fixes, sorry but I forgot to record the error (even after fixing reminaing lib references in generic-builder)

**Comment by angerman** [comment]
@domenkozar I can't comment on `cachix`, as I haven't had a chance to compile that on macOS yet.

I believe this stuff 
https://github.com/NixOS/nixpkgs/blob/2dc883159162f9c1d198092cfd4ccb62904bd36a/pkgs/build-support/bintools-wrapper/ld-wrapper.sh#L207-L210

should be really passed via response files.

And while we do

https://github.com/NixOS/nixpkgs/blob/2dc883159162f9c1d198092cfd4ccb62904bd36a/pkgs/build-support/bintools-wrapper/setup-hook.sh#L26-L28

I think we should just not have `$out/lib` unless it contains proper libraries.

**Comment by domenkozar** [comment]
@angerman fair points. I'm testing on darwin, meanwhile you also need:

```
diff --git a/pkgs/development/haskell-modules/lib.nix b/pkgs/development/haskell-modules/lib.nix
index c9dc7254800..d5e36981ba3 100644
--- a/pkgs/development/haskell-modules/lib.nix
+++ b/pkgs/development/haskell-modules/lib.nix
@@ -234,7 +234,7 @@ rec {
     enableSharedExecutables = false;
     isLibrary = false;
     doHaddock = false;
-    postFixup = "rm -rf $out/lib $out/nix-support $out/share/doc";
+    postFixup = "rm -rf $out/ghc* $out/nix-support $out/share/doc";
   });
 
   /* Build a source distribution tarball instead of using the source files
```

**Comment by angerman** [comment]
@domenkozar right. I've also changed `$out/lib/links` to `$out/lib-links`. That would otherwise inject yet another `$out/lib`.

**Comment by domenkozar** [comment]
@peti I think this can go to testing.

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 1]
```diff
@@ -138,12 +138,12 @@ let
   buildFlagsString = optionalString (buildFlags != []) (" " + concatStringsSep " " buildFlags);
 
   defaultConfigureFlags = [
-    "--verbose" "--prefix=$out" "--libdir=\\$prefix/lib/\\$compiler" "--libsubdir=\\$pkgid"
+    "--verbose" "--prefix=$out" "--libdir=\\$prefix/\\$compiler" "--libsubdir=\\$pkgid"
```
Comment (matthewbauer) [inline pkgs/development/haskell-modules/generic-builder.nix]: This seems weird. Is this a windows things?

File: pkgs/development/compilers/ghc/8.4.3.nix [hunk 2]
```diff
@@ -46,8 +48,7 @@ let
     include mk/flavours/\$(BuildFlavour).mk
     endif
     DYNAMIC_GHC_PROGRAMS = ${if enableShared then "YES" else "NO"}
-  '' + stdenv.lib.optionalString enableIntegerSimple ''
-    INTEGER_LIBRARY = integer-simple
+    INTEGER_LIBRARY = ${if enableIntegerSimple then "integer-simple" else "integer-gmp"}
```
Comment (matthewbauer) [inline pkgs/development/compilers/ghc/8.4.3.nix]: This is definitely a good change! This really confused me with cross-built always pulling in GMP but not ever using it (on android at least).

Comment (matthewbauer) [reply pkgs/development/compilers/ghc/8.4.3.nix]: (might be worth backporting too)

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 3]
```diff
@@ -138,12 +138,12 @@ let
...
     (optionalString enableSeparateDataOutput "--datadir=$data/share/${ghc.name}")
     (optionalString enableSeparateDocOutput "--docdir=${docdir "$doc"}")
     "--with-gcc=$CC" # Clang won't work without that extra information.
     "--package-db=$packageConfDir"
-    (optionalString (enableSharedExecutables && stdenv.isLinux) "--ghc-option=-optl=-Wl,-rpath=$out/lib/${ghc.name}/${pname}-${version}")
+    (optionalString (enableSharedExecutables && stdenv.isLinux) "--ghc-option=-optl=-Wl,-rpath=$out/${ghc.name}/${pname}-${version}")
```
Comment (peti) [inline pkgs/development/haskell-modules/generic-builder.nix]: What is the purpose of this change?

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 4]
```diff
@@ -209,8 +209,8 @@ let
   nativeGhcCommand = "${nativeGhc.targetPrefix}ghc";
 
   buildPkgDb = ghcName: packageConfDir: ''
-    if [ -d "$p/lib/${ghcName}/package.conf.d" ]; then
-      cp -f "$p/lib/${ghcName}/package.conf.d/"*.conf ${packageConfDir}/
+    if [ -d "$p/${ghcName}/package.conf.d" ]; then
+      cp -f "$p/${ghcName}/package.conf.d/"*.conf ${packageConfDir}/
```
Comment (peti) [inline pkgs/development/haskell-modules/generic-builder.nix]: Dito. Why are you changing this?

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 5]
```diff
@@ -294,14 +294,14 @@ stdenv.mkDerivation ({
...
-  # "ln -s $out/lib/links", which tries to recreate the links dir and fails
+  # "ln -s $out/lib-links", which tries to recreate the links dir and fails
   + (optionalString (stdenv.isDarwin && (enableSharedLibraries || enableSharedExecutables)) ''
     # Work around a limit in the macOS Sierra linker on the number of paths
     # referenced by any one dynamic library:
     #
     # Create a local directory with symlinks of the *.dylib (macOS shared
     # libraries) from all the dependencies.
-    local dynamicLinksDir="$out/lib/links"
+    local dynamicLinksDir="$out/lib-links"
```
Comment (peti) [inline pkgs/development/haskell-modules/generic-builder.nix]: Ẁhy?

**Review by peti — CHANGES_REQUESTED** [review]
This PR seems to remove the `lib` directory hierarchy from the generated output, but I don't understand why it does.

**Comment by angerman** [comment]
@matthewbauer, @peti: yes it dropps the `$out/lib`. Having `$out/lib` means that for any haskell dependency it’s `$out/lib` folder will be added to `NIX_LDFLAGS`. This nix feature makes sense when there are `libXXX.ext` or similar libs in that folder and as such linking with `-lXXX` does automatically the right thing.

For Haskell libraries however that makes little sense. First of all there aren’t even any libraries on that directory to beging with. They are in a `ghc...` subdirecotry. Also when linking Haskell libraries you usually use tools that inspect and extract the proper library paths from the package db. Those tools will inject the correct `-L` and `-l` flags. Thus when linking Haskell libs you end up with an almost duplicate set of library search paths, half of which are useless.

Please also refer to the previous discussion here between @domenkozar and me.

**Comment by domenkozar** [comment]
TLDR; this fixes https://github.com/NixOS/nixpkgs/issues/41340 by not polluting LDFLAGS which is unneded for haskell libs.

**Comment by matthewbauer** [comment]
Ok seems reasonable if this fixes the ARG_MAX issue (I wonder if you have tested it with something like stack on macOS). We should definitely be careful because a lot stuff probably depends on things like the package conf dir being ```/lib/ghc-8.4.3/package.conf.d```.

File: pkgs/build-support/cc-wrapper/default.nix [hunk 6]
```diff
@@ -40,7 +40,7 @@ let
   libc_lib = if libc == null then null else getLib libc;
   cc_solib = getLib cc;
   # The wrapper scripts use 'cat' and 'grep', so we may need coreutils.
-  coreutils_bin = if nativeTools then "" else getBin coreutils;
+  coreutils_bin = if (nativeTools || targetPlatform.isWindows) then "" else getBin coreutils;
```
Comment (Ericson2314) [inline pkgs/build-support/cc-wrapper/default.nix]: This `coreutils` for the build platform, not host platform, so why disable it?

File: pkgs/development/compilers/ghc/8.4.3.nix (hunk 2 above)
Comment (Ericson2314) [reply pkgs/development/compilers/ghc/8.4.3.nix]: Yes agreed heartily with both. Great change and please back-port!

File: pkgs/development/compilers/ghc/8.4.3.nix [hunk 7]
```diff
@@ -133,9 +134,11 @@ stdenv.mkDerivation (rec {
   configureFlags = [
     "--datadir=$doc/share/doc/ghc"
     "--with-curses-includes=${ncurses.dev}/include" "--with-curses-libraries=${ncurses.out}/lib"
-  ] ++ stdenv.lib.optional (targetPlatform == hostPlatform && ! enableIntegerSimple) [
+  ] ++ (if (targetPlatform == hostPlatform && !enableIntegerSimple) then [
     "--with-gmp-includes=${gmp.dev}/include" "--with-gmp-libraries=${gmp.out}/lib"
-  ] ++ stdenv.lib.optional (targetPlatform == hostPlatform && hostPlatform.libc != "glibc" && !targetPlatform.isWindows) [
+  ] else [
+    "--with-gmp-includes=${targetPackages.gmp.dev}/include" "--with-gmp-libraries=${targetPackages.gmp.out}/lib"
```
Comment (Ericson2314) [inline pkgs/development/compilers/ghc/8.4.3.nix]: This can just be 
```nix
stdenv.lib.optional (!enableIntegerSimple) [
   "--with-gmp-includes=${targetPackages.gmp.dev}/include" "--with-gmp-libraries=${targetPackages.gmp.out}/lib"
]
```
`targetPackages` = `pkgs` on native.

File: pkgs/development/haskell-modules/generic-builder.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/development/haskell-modules/generic-builder.nix]: This is angerman's change to avoid making cc-wrapper and ld-wrapper inject flags. But please don't do this as part of this PR; even if we end up going this route it should be done separately.

File: pkgs/build-support/cc-wrapper/default.nix (hunk 6 above)
Comment (angerman) [reply pkgs/build-support/cc-wrapper/default.nix]: This ends up trying to build `coreutils` for `mingw32` when cross compiling `gcc`.
And `coreutils` can't be built on `mingw32`.

**Review by peti — CHANGES_REQUESTED** [review]
IMHO, this  goes into the wrong direction. The right solution is not to list Haskell packages in `buildInputs` *at all*. They don't belong there because the generic builder has no clue how to use them anyway. Similarly, one could argue that C libraries required by the Haskell build mustn't be included in `buildInputs` either because it's imperative that Cabal finds them through explicitly configured search paths -- not through some magic `$NIX_XYZ` variables that don't exist outside of the Nix builder.

Arguably, `buildInputs` should be empty except for `build-tools` that can be discovered through `$PATH`.

**Comment by domenkozar** [comment]
@peti this PR fixes several macos haskell packages in pragmatic way, without sacrificing much. It does not touch buildInputs, just renames paths.

While I agree there should be a discussion about how different haskell dependencies are handled and propagated, I'd prefer we either merge these or revert cross compilation fixes that got us here, to be able to build macos again.

**Comment by peti** [comment]
@angerman,

> I agree that haskell lirbaries should not be in `buildInputs`, I do not agree about C libraries though. There is some reasonable expectancy that libraries can be found in the default search paths.

the problem is that Cabal must record the `-I` and `-L` flags required to link the C library into the package's config file, because without that information other packages that try to use the code will fail to compile and/or link. To accomplish that, all those paths must be registered during the configure stage with `--extra-{include,lib}-dirs`. If you do that, then the ability to find those files without giving any explicit search paths -- which `buildInputs` provides -- is unnecessary and arguably even counter-productive.

**Comment by angerman** [comment]
If we are going to pass `--extra-{include,lib}-dirs` for every c library, I agree.

**Comment by peti** [comment]
@angerman,

> If we are going to pass `--extra-{include,lib}-dirs` for every c library, I agree.

we have to, really, because otherwise the generated packages won't work when used outside of the Nix builder.

@domenkozar,

> I'd prefer we either merge these or revert cross compilation fixes that got us here, to be able to build macos again.

I would prefer to revert the changes that broke Darwin support. We can move those patches into a topic branch and have it compiled by Hydra, IMHO, so that @angerman and @Ericson2314 and others can continue their work there and get it in shape for merging it into `master` with reasonable support by the NixOS infrastructure. We do want those improvements after all! But having Darwin support broken for several weeks on end is not a good idea and we have to do something about it.

**Comment by nh2** [comment]
I came here to report that the removal of `/lib` patches in this PR fix the `Argument list too long` issue I am running into (note: on Linux!): https://github.com/NixOS/nixpkgs/issues/41340#issuecomment-405782328

I'm happy to have a workaround now, so thanks @angerman, this came just in time.

I agree with what @peti says. In the long run, we should make sure that all native libs are provided using `--extra-{include,lib}-dirs` so that Cabal knows about them, and putting Haskell libs there doesn't make much sense. Magic discovery at the last step through the linker wrapper is weird and hard to debug.
